incubator/virtualcluster/pkg/vn-agent/server/route.go

```diff
@@ -113,27 +118,103 @@ func (s *Server) InstallHandlers() {
+	if ip == "" {
+		return "",
+			errors.New("cannot get super apiserver ip through env \"KUBERNETES_SERVICE_HOST\"")
 	}
-	tenantName := req.Request.TLS.PeerCertificates[0].Subject.CommonName
-	TranslatePath(req, tenantName)
-
-	klog.V(4).Infof("request after translate %+v", req.Request.URL)
+	httpsPort := os.Getenv("KUBERNETES_SERVICE_PORT_HTTPS")
+	if httpsPort == "" {
+		return "",
+			errors.New("cannot get super apiserver https port through env \"KUBERNETES_SERVICE_PORT_HTTPS\"")
+	}
+	return ip + ":" + httpsPort, nil
+}
 
-	transport := &http.Transport{
-		TLSClientConfig: &tls.Config{
-			InsecureSkipVerify: true,
-			Certificates:       []tls.Certificate{s.config.KubeletClientCert},
-		},
+func (s *Server) proxy(req *restful.Request, resp *restful.Response) {
+	klog.V(4).Infof("request %+v", req.Request.URL)
+	var transport = &http.Transport{}
+
+	if s.config.KubeletClientCert != nil {
+		klog.Info("will forward request to kubelet")
+		// forward request to kubelet
+		req.Request.URL.Host = s.config.KubeletServerHost
+		req.Request.URL.Scheme = "https"
+
+		// there must be a peer certificate in the tls connection
+		if req.Request.TLS == nil || len(req.Request.TLS.PeerCertificates) == 0 {
+			resp.ResponseWriter.WriteHeader(http.StatusForbidden)
+			return
+		}
+		tenantName := req.Request.TLS.PeerCertificates[0].Subject.CommonName
+		TranslatePath(req, tenantName)
+
+		klog.V(4).Infof("request after translate %+v", req.Request.URL)
+
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+				Certificates:       []tls.Certificate{*s.config.KubeletClientCert},
+			},
+		}
+	} else {
+		klog.Info("will forward request to super apiserver")
+		// forward request to super apiserver
+		if req.Request.TLS == nil || len(req.Request.TLS.PeerCertificates) == 0 {
+			resp.ResponseWriter.WriteHeader(http.StatusForbidden)
+			return
+		}
+		// 1. get super master address
+		superHttpsAddr, err := getSuperMasterAddress()
+		if err != nil {
+			klog.Errorf("fail to get apiserver address: %s", err)
+			resp.ResponseWriter.WriteHeader(http.StatusNotFound)
+			resp.ResponseWriter.Write([]byte(err.Error()))
+			return
+		}
+
+		// 2. As we use the sa of the vn-agent, to forward the request, make sure
+		// the sa has the permission to access pod
+		restConfig, err := rest.InClusterConfig()
+		if err != nil {
+			klog.Errorf("fail to sa token or root ca: %s", err)
+			resp.ResponseWriter.WriteHeader(http.StatusNotFound)
+			resp.ResponseWriter.Write([]byte(err.Error()))
+			return
+		}
+
+		tenantName := req.Request.TLS.PeerCertificates[0].Subject.CommonName
+		err = TranslatePathForSuper(req, tenantName)
+		if err != nil {
+			klog.Errorf("fail to translate url path for super master: %s", err)
+			resp.ResponseWriter.WriteHeader(http.StatusNotFound)
+			resp.ResponseWriter.Write([]byte(err.Error()))
+		}
+		klog.V(4).Infof("request after translate %+v", req.Request.URL)
+
```

Comment:
try  `superUrl, err := url.Parse(restConfig.Host)`

Reply:
use `url.Parse` instead of req.Request.URL